src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/StringCache.cs

```diff
@@ -0,0 +1,158 @@
+
+        private bool TryGetValue(string key, out string? value)
+        {
+            if (_hashSet.Contains(new Entry(key)))
+            {
+                var exfiltrator = (ExfiltratingEqualityComparer)_hashSet.Comparer;
+                var val = exfiltrator.LastEqualValue!;
+                if (val.TryGetTarget(out var target))
+                {
+                    value = target!;
+                    return true;
+                }
+                else
+                {
+                    // If one of our strings is out of scope probably tons of them are.
+                    Cleanup();
+                }
+            }
+
+            value = default;
+            return false;
+        }
+
+        /// <summary>
+        /// This is a gross hack to do a sneaky and get the value inside the HashSet out given the lack of any Get operations in netstandard2.0.
+        /// If we ever upgrade to 2.1 delete this and just use the built in.
+        /// </summary>
+        /// <remarks>
+        /// This is fragile on the ordering of the values passed to the EqualityComparer by HashSet.
+        /// If that ever switches we have to react, if it becomes indeterminate we have to abandon this strategy.
+        /// </remarks>
+        private class ExfiltratingEqualityComparer : IEqualityComparer<Entry>
+        {
+            public Entry? LastEqualValue { get; private set; }
+
+            public bool Equals(Entry x, Entry y)
+            {
+                if (x.Equals(y))
+                {
+                    LastEqualValue = x;
+                    return true;
+                }
+                else
+                {
+                    LastEqualValue = null;
+                    return false;
+                }
+            }
+
+            public int GetHashCode(Entry obj)
+            {
+                return obj.TargetHashCode;
+            }
+        }
+
+        private class Entry
+        {
+            public readonly int TargetHashCode;
+
+            private WeakReference<string> _weakRef;
+
+            public Entry(string target)
+            {
+                _weakRef = new WeakReference<string>(target);
+                TargetHashCode = target.GetHashCode();
+            }
+
+            public bool IsAlive => _weakRef.TryGetTarget(out _);
+
+            public bool TryGetTarget(out string target)
+            {
+                return _weakRef.TryGetTarget(out target);
+            }
+
+            public override bool Equals(object obj)
+            {
+                if (!(obj is Entry entry))
+                {
+                    return false;
+                }
+
+                if (TryGetTarget(out var thisTarget) && entry.TryGetTarget(out var entryTarget))
+                {
+                    return thisTarget!.Equals(entryTarget, StringComparison.Ordinal);
+                }
+
+                // If we lost one of the references just compare the HashCodes and hope we don't have any collisions
```

Comment:
If we lost one of the references shouldn't we return `false` in order to ensure we don't get duplicate strings?

Reply:
So the thing here was that in order to be able to Remove items from the HashSet Entries which have lost their Reference need to be able to equate to true, but I'm glad you calls this out because a much better solution to solve only that problem is to do RefEquals.